File: chassis/seed.py
"""Seed content for a fresh demo board.

Realistic FRC-flavored entries so the demo looks alive the moment a tab opens.
Kept separate from store.py to avoid an import cycle.
"""

from datetime import datetime, timezone, timedelta


def _days_ago(n):
    return (datetime.now(timezone.utc) - timedelta(days=n)).isoformat()


def seed_demo_board(new_board, default_tags_fn, uid, now):
    board = new_board("Circuit Breakers", "6328", "FRC", password=None)

    tags = {t["name"]: t["id"] for t in board["tags"]}

    raw = [
        dict(title="Gearbox slipping under load",
             desc="Drivetrain gearbox slipped when pushing against the wall. The set screws "
                  "on the pinion had backed out. Fix: add a flat to the motor shaft and "
                  "thread-lock the set screws. Re-torqued to spec and it held a full match.",
             tagn=["Mechanical"], status="solved", author="maya", pinned=True,
             email="maya.builds@example.com", days=4, progress=100),
        dict(title="Intake rollers stalling on game pieces",
             desc="Compliant wheels grip unevenly and pieces jam at an angle. Trying softer "
                  "durometer wheels on the outer pair, spaced 2mm wider. Still testing.",
             tagn=["Mechanical"], status="in_progress", author="dev", days=1, progress=40),
        dict(title="CAN bus dropping the rear motor controller",
             desc="Random brownouts on the rear-left controller traced to a cold crimp near "
                  "the drivetrain. Re-terminated both ends and added a strain relief loop.",
             tagn=["Electrical"], status="solved", author="priya", pinned=True, days=9, progress=100),
        dict(title="Battery voltage sag during auto",
             desc="Voltage dips below 7V during auto and the controller reboots. Plan: stagger "
                  "the spin-up and add per-subsystem current limits.",
             tagn=["Electrical", "Programming"], status="in_progress", author="sam", days=2, progress=25),
        dict(title="PID tuning for the arm",
             desc="Arm overshoots and rings at the setpoint. Lowered P, added a little D, and a "
                  "gravity feedforward term so it holds position with almost no steady error.",
             tagn=["Programming"], status="solved", author="alex", days=6, progress=100),
        dict(title="Vision lag makes auto-align jittery",
             desc="Pose estimate is about 120ms behind so the robot hunts. Moving the pipeline "
                  "to the coprocessor and feeding latency back into the estimator.",
             tagn=["Programming", "Strategy"], status="in_progress", author="jordan", days=0, progress=55),
        dict(title="Endgame climb sequence",
             desc="Most reliable climb: deploy hooks, drive forward six inches, then retract. "
                  "Doing it in the other order snags the bumper. Captured so next season keeps it.",
             tagn=["Strategy"], status="solved", author="maya", days=12, progress=100),
    ]

    # Insert oldest first so newest ends up on top.
    for r in reversed(raw):
        entry = {
            "id": uid("entry"),
            "title": r["title"],
            "description": r["desc"],
            "tag_ids": [tags[n] for n in r["tagn"] if n in tags],
            "status": r["status"],
            "progress": r["progress"],
            "author": r["author"],
            "contact_email": r.get("email", ""),
            "pinned": r.get("pinned", False),
            "linked_ids": [],
            "comments": [],
            "created_at": _days_ago(r["days"]),
            "updated_at": _days_ago(r["days"]),
        }
        board["entries"].insert(0, entry)

    # A couple of comments on the PID entry to show "corrected over time".
    for e in board["entries"]:
        if e["title"].startswith("PID"):
            e["comments"] = [
                {"id": uid("cmt"), "author": "coach",
                 "text": "Final gains: P=2.4, D=0.15, kG=0.6. Logged on the practice bot.",
                 "created_at": _days_ago(5)},
                {"id": uid("cmt"), "author": "alex",
                 "text": "Bumped D to 0.18 after the heavier intake went on. Smoother now.",
                 "created_at": _days_ago(3)},
            ]

    # A competition date and a few work days so the calendar and stats are populated.
    comp_date = (datetime.now(timezone.utc) + timedelta(days=45)).strftime("%Y-%m-%d")
    board["calendar"]["competition_date"] = comp_date
    for offset in (1, 3, 6, 8):
        d = (datetime.now(timezone.utc) + timedelta(days=offset)).strftime("%Y-%m-%d")
        board["calendar"]["work_days"][d] = {
            "duration": "120", "time": "16:00",
            "timezone": "America/New_York", "location": "School robotics lab",
        }
    return board

File: chassis/src/lib/seed.js
// seed.js
// -----------------------------------------------------------------------------
// First-run demo content. When a brand-new device opens Chassis, we plant a
// sample team with realistic robotics entries so the board "looks like
// something" immediately instead of being an empty shell. Real teams would
// start empty; this is purely so the prototype is explorable on first load.
// -----------------------------------------------------------------------------

import {
  getTeams,
  createTeam,
  ensureDefaultTags,
  getTags,
  addEntry,
  getEntries,
} from './storage'

const DEMO_TEAM = { name: 'Team 6328 Mechanical Advantage', password: 'demo' }

function daysAgo(n) {
  const d = new Date()
  d.setDate(d.getDate() - n)
  return d.toISOString()
}

export function seedIfEmpty() {
  if (getTeams().length > 0) return

  const team = createTeam(DEMO_TEAM)
  ensureDefaultTags(team.id)
  const tags = getTags(team.id)
  const tagId = (name) => tags.find((t) => t.name === name)?.id || null

  if (getEntries(team.id).length > 0) return

  const seedEntries = [
    {
      title: 'Gearbox slipping under load at high current',
      description:
        'Drivetrain gearbox slips when we push against the wall. Found the set screws on the pinion had backed out. Fix: add a flat to the motor shaft and thread-lock the set screws. Re-torqued to spec and it held through a full match.',
      tagName: 'Mechanical',
      status: 'solved',
      author: 'maya',
      pinned: true,
      contactEmail: 'maya.builds@example.com',
      created: 4,
    },
    {
      title: 'Intake rollers stalling on game pieces',
      description:
        'Compliant wheels gripping unevenly — pieces jam at an angle. Trying softer durometer wheels on the outer pair and spacing them 2mm wider. Still testing.',
      tagName: 'Mechanical',
      status: 'in_progress',
      author: 'dev',
      created: 1,
    },
    {
      title: 'CAN bus dropping the rear motor controller',
      description:
        'Random brownouts on the rear-left controller. Traced to a crimp that looked fine but had a cold joint. Lesson: re-crimp anything that flexes near the drivetrain and add a strain relief loop. Solved after re-terminating both ends.',
      tagName: 'Electrical',
      status: 'solved',
      author: 'priya',
      pinned: true,
      created: 9,
    },
    {
      title: 'Battery voltage sag during auto',
      description:
        'Voltage dips below 7V during the auto routine and the RIO reboots. Suspect we are commanding all mechanisms at once. Plan: stagger the spin-up and add current limits per subsystem.',
      tagName: 'Electrical',
      status: 'in_progress',
      author: 'sam',
      created: 2,
    },
    {
      title: 'PID tuning for the arm — oscillation at setpoint',
      description:
        'Arm overshoots and rings. Brought P down, added a touch of D, and a gravity feedforward term so it holds position with almost no steady-state error. Notes on final gains in the comments.',
      tagName: 'Code',
      status: 'solved',
      author: 'alex',
      created: 6,
    },
    {
      title: 'Vision lag making auto-align jittery',
      description:
        'Pose estimate is ~120ms behind, so the robot hunts. Looking at moving the pipeline to the coprocessor and feeding latency back into the estimator. In progress.',
      tagName: 'Code',
      status: 'in_progress',
      author: 'jordan',
      created: 0,
    },
    {
      title: 'Endgame climb sequence — order of operations',
      description:
        'Worked out the climb that scores most reliably: deploy hooks first, then drive forward 6 inches, THEN retract. Doing it in the other order snags the bumper. Captured here so next season does not relearn it the hard way.',
      tagName: 'Strategy',
      status: 'solved',
      author: 'maya',
      created: 12,
    },
  ]

  // Add oldest first so the "unshift" ordering ends up newest-on-top.
  ;[...seedEntries].reverse().forEach((e) => {
    addEntry(team.id, {
      title: e.title,
      description: e.description,
      tagId: tagId(e.tagName),
      status: e.status,
      author: e.author,
      pinned: !!e.pinned,
      contactEmail: e.contactEmail || '',
      createdAt: daysAgo(e.created),
      updatedAt: daysAgo(e.created),
    })
  })

  // Add a couple of comments to the PID entry to show the "correct over time" idea.
  const entries = getEntries(team.id)
  const pid = entries.find((x) => x.title.startsWith('PID tuning'))
  if (pid) {
    pid.comments = [
      {
        id: 'seed_c1',
        author: 'coach',
        text: 'Final gains: P=2.4, I=0, D=0.15, kG=0.6. Logged on the practice bot.',
        createdAt: daysAgo(5),
      },
      {
        id: 'seed_c2',
        author: 'alex',
        text: 'Bumped D to 0.18 after we added the heavier intake. Smoother now.',
        createdAt: daysAgo(3),
      },
    ]
    // persist the comment edit
    const all = getEntries(team.id).map((x) => (x.id === pid.id ? pid : x))
    localStorage.setItem(`chassis.entries.${team.id}`, JSON.stringify(all))
  }
}

export const DEMO_LOGIN = DEMO_TEAM
